images/router/haproxy/shutdown-haproxy

```diff
@@ -0,0 +1,48 @@
+#!/bin/bash
+
+set -euo pipefail
+
+# TERMINATION_MARKER is the path to a file that indicates that this router is
+# shutting down.  It is used as follows:
+#
+# - The reload-haproxy script exits immediately if the marker is present.
+#
+# - This script performs a hard shutdown if the marker is already present, or
+#   else it sets the marker and performs a graceful shutdown.
+TERMINATION_MARKER='/var/lib/haproxy/run/terminating'
+
+# DRAIN_PERIOD is the number of seconds that this script will wait for
+# connections to drain before it sends the SIGUSR1 signal to HAProxy processes.
+DRAIN_PERIOD=30
+
+# GRACE_PERIOD is the number of seconds that this script will wait for HAProxy
+# processes to terminate after it sends the SIGUSR1 signal, before it sends
+# SIGTERM to any remaining HAProxy processes.
+GRACE_PERIOD=30
```

Comment:
Should GRACE_PERIOD and DRAIN_PERIOD be configurable?

Reply:
They easily could be, but I wasn't planning to add any API for these parameters to the ingress controller resource, so I don't know that it would be much use to make them configurable here.

Reply:
The ingress controller is necessarily aware of how to configure the haproxy-router, I'm not sure that implies we have to expose this as public API. I was just suggesting that if we ever need to tweak these values, we might want to do it through an operator release rather than an haproxy-router release. Especially if we end up tweaking these values dynamically in the operator...

Reply:
The stop hook script might even be possible to define in the operator itself rather than here, but that could be a little _too_ coupled?

Reply:
> I was just suggesting that if we ever need to tweak these values, we might want to do it through an operator release rather than an haproxy-router release. Especially if we end up tweaking these values dynamically in the operator...

Hm, that's a good point—the operator might know that one cloud provider needs a shorter or longer drain period than the typical cloud provider, in which case the operator could set a platform-appropriate value.

> The stop hook script might even be possible to define in the operator itself rather than here, but that could be a little _too_ coupled?

You mean with a configmap volume? We could do that, but (1) the reload and shutdown scripts are loosely coupled by the `/var/lib/haproxy/run/terminating` marker, and (2) it would be another piece that could break, so it adds real operational complexity to reduce hypothetical build complexity.

Reply:
Latest push makes `GRACE_PERIOD` and `DRAIN_PERIOD` configurable.